overlord/snapstate/handlers_components.go

```diff
@@ -724,9 +734,45 @@ func (m *SnapManager) doPrepareKernelModulesComponents(t *state.Task, _ *tomb.To
 	// Inject fault during the kernel components setup
 	osutil.MaybeInjectFault("prepare-kernel-components")
 
-	// Make sure we won't be rerun
-	t.SetStatus(state.DoneStatus)
-	return nil
+	var rebootInfo boot.RebootInfo
+	var newInfo *snap.Info
+	prepareRestart := false
+	if err := t.Get("prepare-restart", &prepareRestart); err != nil &&
+		!errors.Is(err, state.ErrNoState) {
+		return err
+	}
+	if prepareRestart {
+		// TODO we have to revert changes in bootloader config/modeenv if an
+		// error happens later in this method. This is not likely as possible
+		// errors after this would happen only due to internal errors or not
+		// being able to write to the filesystem, but still. There is also the
+		// question of what would happen if a restart happens when the boot
+		// configuration has been already written but DoneStatus in the state
+		// has not.
+		cand := sequence.NewRevisionSideState(snapsup.SideInfo, nil)
+		newInfo, err = readInfo(snapsup.InstanceName(), cand.Snap, 0)
+		if err != nil {
+			return err
+		}
+		deviceCtx, err := DeviceCtx(st, t, nil)
+		if err != nil {
+			return err
+		}
+		isUndo := false
+		rebootInfo, err = m.backend.MaybeSetNextBoot(newInfo, deviceCtx, isUndo)
+		if err != nil {
+			return err
+		}
+	}
+
+	if rebootInfo.RebootRequired {
+		return m.finishTaskWithMaybeRestart(t, state.DoneStatus,
+			restartPossibility{info: newInfo, RebootInfo: rebootInfo})
```

Comment:
Could this half of this if statement go in the above if statement's block? Since this shouldn't ever be true if we don't have a valid `rebootInfo`? Then, `rebootInfo` could be scoped to that block.